bliss/models/location_encoder.py

```diff
@@ -305,87 +303,81 @@ def get_n_source_prior_log_prob(self, detection_rate):
-        """Get variational parameters conditioned on number of sources in tile.
+    def encode_for_n_sources_flat(self, var_params_flat: Tensor, tile_n_sources: Tensor):
+        """A wrapper the calls encode_for_n_sources.
+        Deprecated: it's better to call encode_for_n_sources directly,
+        and to always store tiles as a 2D array (i.e., not flattened).
 
         Args:
             var_params_flat: The output of `self.encode(ptiles)`,
                 where the first three dimensions have been flattened.
                 These are the variational parameters in matrix form.
                 Has size `(batch_size x n_tiles_h x n_tiles_w) * d`.
             tile_n_sources:
-                A tensor of the number of sources in each tile.
+                A 1d tensor of the number of sources in each tile.
+
+        Returns:
+            A dictionary where each member has shape
+            `n_ptiles x max_detections x ...`
+        """
+        assert len(tile_n_sources.shape) == 1
+        tile_n_sources = tile_n_sources.unsqueeze(0)
+        ret = self.encode_for_n_sources(var_params_flat, tile_n_sources)
+        return {key: value.squeeze(0) for key, value in ret.items()}
 
-        Raises:
-            ValueError: If the shape of tile_n_sources is not 1 or 2.
+    def encode_for_n_sources(self, var_params_flat: Tensor, tile_n_sources: Tensor):
+        """Get distributional parameters conditioned on number of sources in tile.
+
+        Args:
+            var_params_flat: The output of `self.encode(ptiles)`,
+                where the first three dimensions have been flattened.
+                These are the variational parameters in matrix form.
+                Has size `(batch_size x n_tiles_h x n_tiles_w) * d`.
+            tile_n_sources:
+                A tensor of the number of sources in each tile.
 
         Returns:
-            A dictionary where each member has either shape
+            A dictionary where each member has shape
             `n_samples x n_ptiles x max_detections x ...`
-            or `n_ptiles x max_detections x ...` depending on the shape of `tile_n_sources`.
         """
-
         tile_n_sources = tile_n_sources.clamp(max=self.max_detections)
-        if len(tile_n_sources.shape) == 1:
-            tile_n_sources = tile_n_sources.unsqueeze(0)
-            squeeze = True
-        elif len(tile_n_sources.shape) == 2:
-            squeeze = False
-        else:
-            raise ValueError("tile_n_sources must have shape size 1 or 2")
-
-        assert var_params_flat.shape[0] == tile_n_sources.shape[1]
-        # get probability of params except n_sources
-        # e.g. loc_mean: shape = (n_samples x n_ptiles x max_detections x len(x,y))
-        var_params_for_n_sources = self._get_var_params_for_n_sources(
-            var_params_flat, tile_n_sources
-        )
+        n_ptiles = var_params_flat.size(0)
+        vpf_padded = F.pad(var_params_flat, (0, len(self.variational_params), 0, 0))
```

Comment:
It's not clear to me why `var_parameters_flat` need to be padded. Some of the indices are out-of-bounds though if I don't pad it. The old code was padding `var_parameters_flat` too, just in a less explicit way (which copied the parameters more times)

Reply:
I never understood very well this part of the code, so I appreciate making this more explicit. This could probably be simplified further in a different PR

Reply:
Yes this is a good question and it's a bit subtle. It's been a while that I worked on this code but from what I remember the "off sources" had be indexed separately in the `h` matrix and this is why we had this line: 

```
h = torch.cat((h, torch.zeros(n_ptiles, 1, device=h.device)), dim=1)
```

it's a bit hard to explain but we could discuss in the next meeting. This part of the code is very convoluted so a separate PR will definitely be needed

Reply:
Here is another: https://github.com/prob-ml/bliss/issues/530